Mathlib/Topology/MetricSpace/Bilipschitz.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2024 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import Mathlib.Topology.MetricSpace.Antilipschitz
+import Mathlib.Topology.MetricSpace.Lipschitz
+
+/-! # Bilipschitz equivalence
+
+A common pattern in Mathlib is to replace the topology, uniformity and bornology on a type
+synonym with those of the underlying type.
+
+The most common way to do this is to activate a local instance for something which puts a
+`PseudoMetricSpace` structure on the type synonym, prove that this metric is bilipschitz equivalent
+to the metric on th eunderlying type, and then use this to show that the uniformities and
+bornologies agree, which can then be used with `PseudoMetricSpace.replaceUniformity` or
+`PseudoMetricSpace.replaceBornology`.
+
+With the tooling outside this file, this can be a bit cumbersome, especially when it occurs
+repeatedly, and moreover it can lend itself to abuse of the definitional equality inherant in the
+type synonym. In this file, we make this pattern more convenient by providing lemmas which take
+directly the conditions that the map is bilipschitz, and then prove the relevant equalities.
+Moreover, because there are no type synonyms here, it is necessary to phrase these equalities in
+terms of the induced uniformity and bornology, which means users will need to do the same if they
+choose to use these convenience lemmas. This encourages good hygiene in the development of type
+synonyms.
+-/
+
+open NNReal
+
+section Uniformity
+
+open Uniformity
+
+variable {α β : Type*} [PseudoEMetricSpace α] [PseudoEMetricSpace β]
+variable {K₁ K₂ : ℝ≥0} {f : α → β} (hf₁ : AntilipschitzWith K₁ f) (hf₂ : LipschitzWith K₂ f)
```

Comment:
Should we have `BilipschitzWith` for convenience?

Reply:
maybe, but I don't want to add it yet because I'm not ready to develop the API for it. When we get more use cases, then sure. (There certainly are many)